ios/brave-ios/Sources/PlaylistUI/OEmbedThumbnailFetcher.swift

```diff
@@ -0,0 +1,128 @@
+  static let shared = OEmbedThumbnailFetcher()
+
+  private var providers: [Provider] = []
+  private var hasInitializedProviders: Bool = false
+  private let logger = Logger(
+    subsystem: Bundle.main.bundleIdentifier ?? "com.brave.ios",
+    category: "OEmbedThumbnailFetcher"
+  )
+
+  /// Loads the providers file
+  private func loadProviders() async {
+    defer { hasInitializedProviders = true }
+    if hasInitializedProviders {
+      return
+    }
+    guard let path = Bundle.module.path(forResource: "oembed_providers", ofType: "json") else {
+      return
+    }
+    do {
+      let data = try Data(contentsOf: URL(filePath: path))
+      let decoder = JSONDecoder()
+      providers = try decoder.decode([FailableDecodable<Provider>].self, from: data).compactMap {
+        $0.wrappedValue
+      }
+    } catch {
+      logger.error("Failed to load oembed providers: \(error.localizedDescription)")
+    }
+  }
+
+  private func endpoint(for url: URL) async -> Provider.Endpoint? {
+    await loadProviders()
+
+    let domain = url.baseDomain
+    guard let provider = providers.first(where: { $0.providerURL?.baseDomain == domain }) else {
+      return nil
+    }
+
+    for endpoint in provider.endpoints {
+      guard let schemes = endpoint.schemes else { continue }
+      // Each scheme will be a URL that may contain wildcard (`*`) entries in it to match against
+      // many different pages on the site that may vend said embedded content.
+      //
+      // We can use NSPredicate's `LIKE` keyword here to match wildcards properly, but unfortunately
+      // we need to drop down to `NSArray` to use the `filtered(using:)` method to apply the
+      // predicate.
+      let predicate = NSCompoundPredicate(
+        type: .or,
+        subpredicates: schemes.map {
+          NSPredicate(format: "self LIKE %@", $0)
+        }
+      )
+      if !([url.absoluteString] as NSArray).filtered(using: predicate).isEmpty {
+        return endpoint
+      }
+    }
+    return nil
+  }
+
+  /// Fetch a oembed thumbnail for a given URL, or return `nil` if the provider isn't supported
+  /// or an error occured
+  func oembedThumbnail(for url: URL) async -> UIImage? {
+    guard let endpoint = await endpoint(for: url),
+      let providerURL = endpoint.url,
+      var components = URLComponents(url: providerURL, resolvingAgainstBaseURL: false)
+    else {
+      return nil
+    }
+    components.queryItems = [
+      .init(name: "url", value: url.absoluteString),
+      .init(name: "format", value: "json"),
+    ]
+    guard let requestURL = components.url else { return nil }
+    let request = URLRequest(url: requestURL)
+    do {
+      let (data, _) = try await URLSession.shared.data(for: request)
+      let response = try JSONDecoder().decode(ProviderResponse.self, from: data)
+      guard let thumbnailURL = response.thumbnailURL else {
+        return nil
+      }
+      let (thumbnailData, _) = try await URLSession.shared.data(from: thumbnailURL)
```

Comment:
there is a lot of http urls in the json. how does HTTPS autoupgrade work with it?

Reply:
It would not participate since none of this happens inside of the browser web views.

Reply:
When I go to http://animoto.com/ it redirects me to https://animoto.com/. If I add a video to playlist from there Brave will send the request to http, because we get the base domain from the url and look it up in the json, which has this:
```
{
        "provider_name": "Animoto",
        "provider_url": "http://animoto.com/",
        "endpoints": [
            {
                "schemes": [
                    "http://animoto.com/play/*"
                ],
                "url": "http://animoto.com/oembeds/create"
            }
        ]
    }
```

That doesn't look good to me.

Reply:
I'd say we should override the scheme and use the one from the tab

Reply:
I mean if we want I can just force https for all oembed calls, if they fail there's always a fallback anyways so its not a huge deal

Reply:
Sounds good!

Reply:
Added explicit scheme set here: https://github.com/brave/brave-core/pull/24163/commits/3f807622be01360119091a080b836183d93e801c#diff-88e28104cda111e833c2efb99205b1d4b07ef428c1a294e0e8083dfc19ab7959R80